021-server-defined-attributes.md

```diff
@@ -0,0 +1,120 @@
+# Optional server-defined attributes
+
+* Creator: Khushboo Verma
+* Relevant Issues: N/A
+
+- [RFC Template](#rfc-template)
+  - [Summary](#summary)
+  - [Resources](#resources)
+  - [Implementation](#implementation)
+    - [API Changes](#api-changes)
+    - [Workers / Commands](#workers--commands)
+    - [Supporting Libraries](#supporting-libraries)
+    - [Data Structures](#data-structures)
+    - [SDKs](#sdks)
+    - [Breaking Changes](#breaking-changes)
+    - [Documentation & Content](#documentation--content)
+  - [Reliability](#reliability)
+    - [Security](#security)
+    - [Scaling](#scaling)
+    - [Benchmarks](#benchmarks)
+    - [Tests (UI, Unit, E2E)](#tests-ui-unit-e2e)
+  - [Open Questions](#open-questions)
+  - [Future Possibilities](#future-possibilities)
+
+## Summary
+<!-- Describe the problem we want to solve and suggested solution in a few paragraphs -->
+This RFC proposes the addition of optional server-defined attributes in all the project collections that can only be added by server and cannot be modified by the client. The user can add them to each collection through the API or console. Some useful sttributes can be `createdBy`, `updatedBy`, `ip` etc.
+
+## Resources
+<!-- List all the resources used for writing this RFC -->
+NA
+
+## Implementation
+
+<!-- Write an overview to explain the suggested implementation -->
+#### Console Changes
+We need to add an option in the `create attribute` settings dropdown to add a certain server-defined attribute from one of the existing options like `createdBy`.
+
+### API Changes
+<!-- Do we need new API endpoints? List and describe them and their API signatures -->
+We need 2 new API endpoints in `databases.php` for creating and deleting server-defined attributes.
+- POST `/v1/databases/:databaseId/collections/:collectionId/attributes/server`
+- DELETE `/v1/databases/:databaseId/collections/:collectionId/attributes/server/:key`
+
+We can allow users to create indexes for new attributes from the `createIndex` endpoint.
+
+We need to add a check in `createDocument` and `updateDocument` endpoints to populate the values of these attributes if they exist for client-side requests.
+
+###  Workers / Commands
+<!-- Do we need new workers or commands for this feature? List and describe them and their API signatures -->
+No new workers are anticipated.
+
+###  Supporting Libraries
+<!-- Do we need new libraries for this feature? Mention which, define the file structure, and different interfaces -->
+No new changes required in libraries.
```

Comment:
Are you sure? Database library already auto-provides updatedAt, createdAt. These would also become server attributes, and it would feel weird to have some in Appwrite, and some in library.

Reply:
Yes, resolving as discussed.